### PR TITLE
reef: cephfs: implement snapdiff

### PR DESCRIPTION
**Pull request by ifed01**
Backport of https://github.com/ceph/ceph/pull/43546
Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by joscollin — APPROVED** [review]

**Review by vshankar — CHANGES_REQUESTED** [review]
@ifed01 PTAL

File: src/mds/Server.cc [hunk 1]
```diff
@@ -11457,6 +11481,159 @@ void Server::_renamesnap_finish(MDRequestRef& mdr, CInode *diri, snapid_t snapid
...
+
+  SnapRealm* realm = diri->find_snaprealm();
+
+  unsigned max = req->head.args.snapdiff.max_entries;
+  if (!max)
+    max = dir->get_num_any();  // whatever, something big.
+  unsigned max_bytes = req->head.args.snapdiff.max_bytes;
+  if (!max_bytes)
+    // make sure at least one item can be encoded
+    max_bytes = (512 << 10) + g_conf()->mds_max_xattr_pairs_size;
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/mds/Server.cc:11604]: false alarm -- probably build script bug. rebuilding.

**Review by lxbsz — APPROVED** [review]
Checked all the 15 test failures and all are not related to this PR:

https://pulpito.ceph.com/vshankar-2023-10-09_16:48:27-fs-wip-vshankar-testing-reef-20231009.131610-testing-default-smithi/

LGTM.

**Comment by vshankar** [comment]
> @vshankar - can we go ahead with merging?

@yuriw merges backport normally - I'm sure he is on it since @lxbsz approved the change.

**Comment by vshankar** [comment]
> > @vshankar - can we go ahead with merging?
> 
> @yuriw merges backport normally - I'm sure he is on it since @lxbsz approved the change.

Spoke to @yuriw - this change (along with others in the test batch) will be merged today.

**Comment by yuriw** [comment]
@vshankar you need to remove your block here

**Comment by vshankar** [comment]
> @vshankar you need to remove your block here

Done.

**Pull request merged**